demos/companion-app/src/components/fund-arcade-account.tsx

```diff
@@ -0,0 +1,134 @@
+"use client";
+import { useAccount, useContractWrite } from "@starknet-react/core";
+import { useState } from "react";
+import { CallData, stark, constants, uint256, hash } from "starknet";
+import { z } from "zod";
+
+import { Input } from "./ui/input";
+import { useForm } from "react-hook-form";
+import { Label } from "./ui/label";
+import { zodResolver } from "@hookform/resolvers/zod";
+import { Button } from "./ui/button";
+import { Card, CardContent } from "./ui/card";
+
+interface Props {
+  pk: string;
+}
+const TOKEN_CLASSS_HASH =
+  "0x0278da3cbbc2105fc3ac1206a630357c5a4666020f15fed02c892ac5d856d8ef";
+
+const schema = z.object({
+  initialSupply: z.number().min(0),
+});
+
+export const FundArcadeAccount = ({ pk }: Props) => {
+  const { address } = useAccount();
+
+  const {
+    register,
+    handleSubmit,
+    watch,
+    formState: { errors },
+    reset: resetForm,
+  } = useForm<z.infer<typeof schema>>({
+    resolver: zodResolver(schema),
+    defaultValues: { initialSupply: 10_000_000_000 },
+  });
+
+  const { writeAsync, reset: resetWrite } = useContractWrite({});
+
+  const [deployedToken, setDeployedToken] = useState<
+    { address: string; tx: string } | undefined
+  >(undefined);
+
+  const amount = watch("initialSupply");
+  const salt = stark.randomAddress();
+
+  const deployToken = async (data: z.infer<typeof schema>) => {
+    if (!address) return;
+
+    const unique = 0;
+
+    const constructorCalldata = CallData.compile([
+      address,
+      address,
+      uint256.bnToUint256(BigInt(data.initialSupply)),
+    ]);
+
+    const tokenAddress = hash.calculateContractAddressFromHash(
+      salt,
+      TOKEN_CLASSS_HASH,
+      constructorCalldata,
+      unique
+    );
+
+    const deploy = {
```

Comment:
The good news is that now we have [a factory contract here](https://testnet.starkscan.co/contract/0x06c595dced9a22879e55f11f909d334ddfb976b4a65da0122778c5ae909a0e33#read-write-contract).

Notice two methods:

 * `compute_address(salt, pk, master)` (read): returns the future address of the deployed account. Use this to 1) display the deployed address to the user 2) know where to send the funds.
 * `deploy(salt, pk, master)` (write): use this to deploy the account

So the final call will be:

 * transfer from master to arcade account
 * deploy arcade account
 * (in the future) authorize the demo app